### ISSUE TITLE
Strange behavior in AstRef.__bool__() in Python API?

### ISSUE THREAD

**Issue by zwegner** [Z3Prover/z3#2806]
While writing a model for x86 instructions, I needed to check if a certain expression is statically determinable, to resolve a branch. I was getting a weird result where an equality check was being resolved to `False`, when it should not be determinable. I traced the problem to this snippet of code (https://github.com/Z3Prover/z3/blob/master/src/api/python/z3/z3.py#L335-L343):

```python
    def __bool__(self):
        if is_true(self):
            return True
        elif is_false(self):
            return False
        elif is_eq(self) and self.num_args() == 2:
           return self.arg(0).eq(self.arg(1))
        else:
            raise Z3Exception("Symbolic expressions cannot be cast to concrete Boolean values.")
```

The third condition means that an equality check between two distinct expressions will always yield `False`, when I'd expect it to raise an exception. 

I can see this check being useful for returning `True` for an expression `x == x`, but the inverse is not necessarily true. As a side note, if `x >= x` and `x <= x` could be resolved to True as well, that would save me some additional hacks in this code...
 
Here's a short reproduction case:
```python
import z3
a = z3.BitVec('a', 1)
b = z3.BitVec('b', 1)
eq = (a == b)
assert not z3.is_true(eq) and not z3.is_false(eq)
print(bool(eq))  # <-- False?
```

**Comment by NikolajBjorner** [comment]
First some basic background, which I think you already get:

The operator == is an _expression_ builder. It builds the equality expression, un-simplified.
If you want a syntactic equality check use the a.eq(b) operator. It produces either True or False depending on whether the expressions are syntactically equal. For example if x is defined as an Integer constant (x = Int('x')), then (x + 1).eq(1 + x) is False and (x + 1).eq(x + 1) is True.

The bool operator then is defined as a slight extension to eq. It returns True if it can be established, very cheaply, that the expression is semantically equivalent to True. 

Now to your post:
It is indeed odd that it raises an exception for expressions like Ite(a == b, True, False), but not a == b.
This could be said to be arbitrary. 
I believe the __bool__ operator is rarely used, if at all. It could be redefined as appropriate, but likely there is a tutorial example that uses this feature, e.g.,

https://notebooks.azure.com/nbjorner/projects/z3examples

For your consumption you could rely on a check, such as, simplify(eq).eq(BoolVar(True)).
There is a catch. Simplify could be too expensive for your purposes. 
It doesn't cache simplifications between calls so the overhead of using this feature may
outweigh the overhead of performing check_sat() calls.

The options therefore seem to be:
1. You pre-process on your own as appropriate in a tuned and incremental way that fits your application and don't rely on Z3 to do shortcuts.
2. You can propose an interaction model that that is suitable for applications for yours.
    For example: simplification rewriting that caches between calls.
3. Maybe you could propose something more useful.

**Comment by zwegner** [comment]
Thanks for the reply!

> First some basic background, which I think you already get:
> [snip]

I knew the part about expressions, but didn't know `.eq()` is just a syntactic check--I'm still new to Z3. I didn't expect too much magic from `bool()`, I am running it after simplifying, which I do expect magic from :)

> For your consumption you could rely on a check, such as, simplify(eq).eq(BoolVar(True)).

Yes, that is almost what I ended up with. Before I was checking `bool(simplify(expr))` and using an exception handler to know when the value is statically unknown. Now I check `is_true` and `is_false`, which I think is just as clear. I just found this particular part of `bool()` pretty surprising/unusual, and it took a little bit to debug.

> There is a catch. Simplify could be too expensive for your purposes.
> It doesn't cache simplifications between calls so the overhead of using this feature may
> outweigh the overhead of performing check_sat() calls.
> 
> The options therefore seem to be:
> 
>     1. You pre-process on your own as appropriate in a tuned and incremental way that fits your application and don't rely on Z3 to do shortcuts.
> 
>     2. You can propose an interaction model that that is suitable for applications for yours.
>        For example: simplification rewriting that caches between calls.
> 
>     3. Maybe you could propose something more useful.

It's good to know that `simplify()` is expensive and not cached--that will definitely affect how I use it.

I could cache things myself, but this would only capture the root-level expressions. I'd expect deeper caching on sub-expressions to be much more effective (sometimes even within a single `simplify()`), at least if expressions can be canonicalized quickly. But that's probably a lot of work!

I'm not sure how much pre-processing I want to do: this project is still a toy at this point, and I want to keep it as simple as possible. BTW, it is now published as open source: https://github.com/zwegner/x86-sat
